rwkv/chat_with_bot.py

```diff
@@ -47,44 +115,168 @@
 print('Loading RWKV model')
 model = rwkv_cpp_model.RWKVModel(library, args.model_path)
 
-prompt_tokens = tokenizer.encode(prompt).ids
+prompt_tokens = tokenizer.encode(init_prompt).ids
 prompt_token_count = len(prompt_tokens)
 print(f'Processing {prompt_token_count} prompt tokens, may take a while')
 
-logits, state = None, None
+
+########################################################################################################
+def run_rnn(tokens):
+    global model_tokens, model_state, logits
+
+    tokens = [int(x) for x in tokens]
+    model_tokens += tokens
+
+    for token in tokens:
+        logits, model_state = model.eval(token, model_state, model_state, logits)
+    
+    return logits
+
+
+all_state = {}
+def save_all_stat(srv, name, last_out):
+    n = f'{name}_{srv}'
+    all_state[n] = {}
+    all_state[n]['logits'] = copy.deepcopy(last_out)
+    all_state[n]['rnn'] = copy.deepcopy(model_state)
+    all_state[n]['token'] = copy.deepcopy(model_tokens)
+
+def load_all_stat(srv, name):
+    global model_tokens, model_state
+    n = f'{name}_{srv}'
+    model_state = copy.deepcopy(all_state[n]['rnn'])
+    model_tokens = copy.deepcopy(all_state[n]['token'])
+    return copy.deepcopy(all_state[n]['logits'])
+
+########################################################################################################
+
+model_tokens = []
+logits, model_state = None, None
 
 for token in prompt_tokens:
-    logits, state = model.eval(token, state, state, logits)
+    logits, model_state = model.eval(token, model_state, model_state, logits)
+    model_tokens.append(token)
 
+save_all_stat('', 'chat_init', logits)
 print('\nChat initialized! Write something and press Enter.')
 
+srv_list = ['dummy_server']
+for s in srv_list:
+    save_all_stat(s, 'chat', logits)
+
+
+def reply_msg(msg):
+    print(f'{bot}{interface} {msg}\n')
+
+
 while True:
     # Read user input
-    print('> ', end='')
-    user_input = sys.stdin.readline()
+    user_input = input(f'> {user}{interface} ')
+    msg = user_input.replace('\\n','\n').strip()
 
-    # Process the input
-    new_tokens = tokenizer.encode('\n' + user_message_prefix + ' ' + user_input + '\n' + bot_message_prefix).ids
+    srv = 'dummy_server'
 
-    for token in new_tokens:
-        logits, state = model.eval(token, state, state, logits)
+    if msg == '+reset':
+        logits = load_all_stat('', 'chat_init')
+        save_all_stat(srv, 'chat', logits)
+        reply_msg("Chat reset.")
+        continue
+    elif msg[:5].lower() == '+gen ' or msg[:3].lower() == '+i ' or msg[:4].lower() == '+qa ' or msg[:4].lower() == '+qq ' or msg.lower() == '+++' or msg.lower() == '++':
```

Comment:
I think all commands need to be documented, somewhere at the start of the file. It's hard to determine from code alone what each command does.